cloudinit/sources/DataSourceAzure.py

```diff
@@ -596,25 +596,32 @@ def _poll_imds(self):
         return_val = None
 
         def exc_cb(msg, exception):
-            if isinstance(exception, UrlError) and exception.code == 404:
-                if self.imds_poll_counter == self.imds_logging_threshold:
-                    # Reducing the logging frequency as we are polling IMDS
-                    self.imds_logging_threshold *= 2
-                    LOG.debug("Call to IMDS with arguments %s failed "
-                              "with status code %s after %s retries",
-                              msg, exception.code, self.imds_poll_counter)
-                    LOG.debug("Backing off logging threshold for the same "
-                              "exception to %d", self.imds_logging_threshold)
-                self.imds_poll_counter += 1
-                return True
-
-            # If we get an exception while trying to call IMDS, we
-            # call DHCP and setup the ephemeral network to acquire the new IP.
-            LOG.debug("Call to IMDS with arguments %s failed  with "
-                      "status code %s", msg, exception.code)
-            report_diagnostic_event("polling IMDS failed with exception %s"
-                                    % exception.code)
-            return False
+            if isinstance(exception, UrlError):
+                if exception.code in (404, 410):
+                    if self.imds_poll_counter == self.imds_logging_threshold:
+                        # Reducing the logging frequency as we are polling IMDS
+                        self.imds_logging_threshold *= 2
+                        LOG.debug("Call to IMDS with arguments %s failed "
+                                  "with status code %s after %s retries",
+                                  msg, exception.code, self.imds_poll_counter)
+                        LOG.debug("Backing off logging threshold for the same "
+                                  "exception to %d",
+                                  self.imds_logging_threshold)
+                    self.imds_poll_counter += 1
+                    return True
+                else:
+                    url_error = UrlError(exception)
```

Comment:
If isinstance(exeption, URLError), it seems like we are potentially re-wrapping any UrlError exception that doesn't have a code in (404, 410).

Shouldn't we check the instance type of exception before wrapping it again?
It'd be nice to have a unittest getting coverage through this error path. (either raising a 500 URLError exception.code or an unexpected RuntimeError)

Reply:
The url_helper.readurl method always wraps the exception in URLError before it passes it back to the exception call back method. So we should be good on this regard.

Reply:
Since we are only down this path   "if isinstance(exception, UrlError):"  shouldn't we be able to just do the following assignment?
```suggestion
                    url_error = exception
```